invenio_rdm_migrator/state.py

```diff
@@ -0,0 +1,248 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-RDM-Migrator is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Base state module."""
+
+from abc import ABC, abstractclassmethod
+from pathlib import Path
+from uuid import UUID
+
+from sqlalchemy import (
+    Column,
+    Integer,
+    MetaData,
+    String,
+    Table,
+    create_engine,
+    insert,
+    select,
+    update,
+)
+from sqlalchemy_utils.types import JSONType, UUIDType
+
+from .logging import Logger
+from .utils import ts
+
+
+class State:
+    """Migration state."""
+
+    def __init__(self, dbpath, validators=None):
+        """Constructor.
+
+        :param dbpath: path to sqlite db file.
+        """
+        # for extra validation that cannot be achieved with SQL constraints
+        self.validators = validators or {}
+        self.dbpath = Path(dbpath) / "state.db"
+        self.mem_eng = create_engine(f"sqlite:///:memory:")
+        self.metadata = MetaData()
+        self._initialize_db(self.metadata)  # bind tables to metadata
+
+        start = ts(iso=False)
+        logger = Logger.get_logger()
+        logger.info(f"Loading state from {self.dbpath} {ts()}")
+
+        disk_eng = create_engine(f"sqlite:///{self.dbpath}")
+        self.metadata.create_all(disk_eng)  # create tables if not exist
+
+        self._copy_db(disk_eng, self.mem_eng)  # dump state into in-memory sqlite db
+
+        end = ts(iso=False)
+        logger.info(f"Finished loading state {ts()}")
+        logger.info(f"State loading took {end-start} seconds.")
+
+    def save(self):
+        """Save the current in-memory state to disk.
+
+        It will overwrite the loaded db file.
+        """
+        logger = Logger.get_logger()
+        start = ts(iso=False)
+
+        # move current db state to backup
+        disk_file = Path(self.dbpath)
+        logger.info(f"Dumping state to {disk_file} {ts()}")
+        backup_file = Path(self.dbpath.name + ".backup")
+        if disk_file.exists():
+            disk_file.rename(backup_file)
```

Comment:
Maybe is worth discussing but as I understood it, there is no need to namespace this. Most of the tables are independent per string, except e.g. parents which needs both. While it is true that we can dump the state after each stream, the final/current state will always be that after the last stream. You want to namespace in case you want to re-take it from a middle stream?

Reply:
> You want to namespace in case you want to re-take it from a middle stream?

Poorly worded on my side, but yes, what I meant was preserving the state after each stream finishes. 

What I had in mind is the following streams running scenario:

1. Run Users
2. Run Communities
3. Run Records
4. Run Drafts

We find a bug in the Records stream and want to run again everything from that point. We could restore the state dumped after Communities finished, and go on from there.